File: sample/sample-java/SampleMonitoring.java
//---------------------------------------------------------------------------
//
// TSDuck sample Java application running a chain of plugins on the long
// run with CPU and memory monitoring.
//
//----------------------------------------------------------------------------

import java.io.File;

import io.tsduck.AsyncReport;
import io.tsduck.Info;
import io.tsduck.Report;
import io.tsduck.SystemMonitor;
import io.tsduck.TSProcessor;

public class SampleMonitoring {

    private static final String URL = "https://tsduck.io/streams/france-dttv/tnt-uhf30-546MHz-2019-01-22.ts";

    /**
     * Main program.
     * @param args Command line arguments.
     */
    public static void main(String[] args) throws Exception {

        System.out.printf("TSDuck version: %s\n", Info.version());

        // Create an asynchronous report to log multi-threaded messages.
        AsyncReport rep = new AsyncReport();
        rep.setMaxSeverity(Report.Verbose);

        // Create and start a background system monitor.
        SystemMonitor mon = new SystemMonitor(rep);
        mon.start();

        // Build a temporary file name to download a real TS file.
        File tsfile = File.createTempFile("temp", ".ts", null);
        tsfile.deleteOnExit();

        // First phase: Download the TS file:
        System.out.printf("Downloading %s to %s ...\n", URL, tsfile.getPath());

        TSProcessor tsp = new TSProcessor(rep);
        tsp.input = new String[] {"http", URL};
        tsp.output = new String[] {"file", tsfile.getPath()};
        tsp.start();
        tsp.waitForTermination();
        tsp.delete();

        // Second phase: Play the file at regulated speed several times.
        // Must use another instance of TSProcessor.
        System.out.printf("Playing %s ...\n", tsfile.getPath());

        tsp = new TSProcessor(rep);
        tsp.input = new String[] {"file", tsfile.getPath(), "--repeat", "2"};
        tsp.plugins = new String[][] {{"regulate"}};
        tsp.output = new String[] {"drop"};
        tsp.start();
        tsp.waitForTermination();
        tsp.delete();

        // Terminate the system monitor.
        mon.stop();
        mon.waitForTermination();
        mon.delete();

        // Terminate the asynchronous report.
        rep.terminate();
        rep.delete();
    }
}

File: sample/sample-python/sample-monitoring.py
#!/usr/bin/env python
#----------------------------------------------------------------------------
#
# TSDuck sample Python application running a chain of plugins on the long
# run with CPU and memory monitoring.
#
#----------------------------------------------------------------------------

import tsduck
import os
import tempfile

# Create an asynchronous report to log multi-threaded messages.
rep = tsduck.AsyncReport(severity = tsduck.Report.Verbose)

# Create and start a background system monitor.
mon = tsduck.SystemMonitor(rep)
mon.start()

# Build a temporary file name to download a real TS file.
url = "https://tsduck.io/streams/france-dttv/tnt-uhf30-546MHz-2019-01-22.ts"
tsfile = tempfile.gettempdir() + os.path.sep + tempfile.gettempprefix() + str(os.getpid()) + ".ts"

# First phase: Download the TS file:
print("Downloading %s to %s ..." % (url, tsfile))

tsp = tsduck.TSProcessor(rep)
tsp.input = ['http', url]
tsp.output = ['file', tsfile]
tsp.start()
tsp.waitForTermination()
tsp.delete()

# Second phase: Play the file at regulated speed several times.
# Must use another instance of tsduck.TSProcessor.
print("Playing %s ..." % (tsfile))

tsp = tsduck.TSProcessor(rep)
tsp.input = ['file', tsfile, '--repeat', '2']
tsp.plugins = [ ['regulate'] ]
tsp.output = ['drop']

tsp.start()
tsp.waitForTermination()
tsp.delete()

# Terminate the system monitor.
mon.stop()
mon.waitForTermination()
mon.delete()

# Terminate the asynchronous report.
rep.terminate()
rep.delete()

# Delete temporary TS file.
os.remove(tsfile)
